File: src/c++/network_flow/bipartite_matching/11376/main.cc
/* Date : 2019.5.9
 * Author : seungyeon choi
 * 11376 (열혈강호2)
 */

#include <cstdio>
#include <vector>

using namespace std;

const int MAX = 1001;

int N, M;
bool adj[MAX][MAX];
vector<int> cow, house;
vector<int> visit;

bool dfs(int i) {
  if (visit[i] > 0) return false;
  visit[i]++;
  for (int j = 0; j < M; j++) {
    if (adj[i][j]) {
      if (house[j] == -1 || dfs(house[j])) {
        cow[i] = j;
        house[j] = i;
        return true;
      }
    }
  }
  return false;
}

int bmatch() {
  cow = vector<int>(N, -1);
  house = vector<int>(M, -1);
  int ans = 0;
  for (int i = 0; i < N; i++) {
    visit = vector<int>(N, -1);
    if (dfs(i))
      ans++;
    if (dfs(i))
      ans++;
  }
  return ans;
}

int main() {
  int num, num2 = 0;
  scanf("%d %d", &N, &M);
  for (int i = 0; i < N; i++) {
    scanf("%d", &num);
    for (int j = 0; j < num; j++) {
      scanf("%d", &num2);
      adj[i][num2 - 1] = true;
    }
  }

  printf("%d\n", bmatch());

  return 0;
}

File: src/python/network_flow/bipartite_matching/11376/main.py
# Date : 2017.11.10
# Author : seungyeon choi
# 11376 (열혈강호2)

import sys
ans = 0
v = 1
N, M = map(int, sys.stdin.readline().split())
adj = [[] for _ in range(N)]
cow = [-1 for _ in range(N)]
house = [-1 for _ in range(M)]
visit = [0 for _ in range(N)]
for i in range(N):
    S =  list(map(int, sys.stdin.readline().split()))
    for j in range(1, len(S)):
        adj[i].append(S[j]-1)
for j in range(N):
    def dfs(i):
        global v
        if visit[i] == v : return False
        visit[i] = v
        for j in range(len(adj[i])):
            if house[adj[i][j]] == -1 or dfs(house[adj[i][j]]):
                cow[i] = adj[i][j]
                house[adj[i][j]] = i
                return True
        return False
    v += 1
    if dfs(j): ans += 1
    v += 1
    if dfs(j): ans += 1
print (ans)
